src/edu/psu/sweng500/team8/play/GameSession.java

```diff
@@ -55,24 +53,61 @@ public Board getGameBoard() {
 	 * If we want to use return object, we maybe able to use Enum to show previously defined exception cases.
 	 * 
 	 *
-	 * @param cellCoordinates
+	 * @param cell
 	 * @param number
 	 */
-	public void enterNumber(CellCoordinates cellCoordinates, int number) {
-
+	public void enterNumber(Cell currentCell, int number) {
+		
 		/* keep track of the last action*/
 		SudokuAction sudokuAction = new SudokuAction(new CellGrid(board.getCellGrid()));
 		
-		/* updating number */
-		Cell currentCell = board.getCell(cellCoordinates.getRowIndex(), cellCoordinates.getColumnIndex());
-		currentCell.setNumber(number);
+		boolean isDelete = false;
 		
-		actionManager.addAction(sudokuAction);
+		if (number == 0) {
+			isDelete = true;			
+		}
+		
+		if (isDelete) {
+			currentCell.clearNumber();
+		} else {
+			currentCell.setNumber(number);	
+		}
+		
+		/* FIXME - PencilMarkManager updating/wiping out PencilMark number matching entered number */
+		if (!isDelete) {
+			for (int i = 0; i < 9; i++) {
+				for (int j = 0; j < 9; j++) {
+					Cell eachCell = this.getGameBoard().getCellGrid().getCell(i, j);
+					enterPencilMark(eachCell, number, false);
+				}
+			}			
+		}
 		
-		/* TODO - PencilMarkManager updating/wiping out PencilMark number matching entered number */
+		actionManager.addAction(sudokuAction);
 	}
+
 	
 	/**
+	 * TODO implement
+	 * @param currentCell
+	 * @param number
+	 */
+	public void enterPencilMark(Cell currentCell, int number, boolean isEnter) {
+		
+		/* keep track of the last action*/
+		SudokuAction sudokuAction = new SudokuAction(new CellGrid(board.getCellGrid()));

```

Comment:
We might need to rethink how this is implemented if this becomes a memory issue. Storing a copy of all 81 cells for each user action seems like a lot of overhead.